framework/src/utils/BicubicSplineInterpolation.C

```diff
@@ -0,0 +1,226 @@
+  else
+    for (int i = 0; i < _y.size(); ++i)
+      if (_y[i].size() != n)
+        mooseError("y column dimension does not match the size of x2.");
+
+  if (_yx11.empty())
+    _yx11.resize(n, 1e30);
+  else if (_yx11.size() != n)
+    mooseError("The length of the vectors holding the first derivatives of y with respect to x1 must match the length of x2.");
+
+  if (_yx1n.empty())
+    _yx1n.resize(n, 1e30);
+  else if (_yx1n.size() != n)
+    mooseError("The length of the vectors holding the first derivatives of y with respect to x1 must match the length of x2.");
+
+  if (_yx21.empty())
+    _yx21.resize(m, 1e30);
+  else if (_yx21.size() != m)
+    mooseError("The length of the vectors holding the first derivatives of y with respect to x2 must match the length of x1.");
+
+  if (_yx2n.empty())
+    _yx2n.resize(m, 1e30);
+  else if (_yx2n.size() != m)
+    mooseError("The length of the vectors holding the first derivatives of y with respect to x2 must match the length of x1.");
+}
+
+void
+BicubicSplineInterpolation::constructRowSplineSecondDerivativeTable()
+{
+  int m = _x1.size();
+  _y2_rows.resize(m);
+
+  if (_yx21.empty())
+    for (int j = 0; j < m; ++j)
+      spline(_x2, _y[j], _y2_rows[j]);
+
+  else
+    for (int j = 0; j < m; ++j)
+      spline(_x2, _y[j], _y2_rows[j], _yx21[j], _yx2n[j]);
+}
+
+void
+BicubicSplineInterpolation::constructColumnSplineSecondDerivativeTable()
+{
+  int n = _x2.size();
+  _y2_columns.resize(n);
+
+  if (_yx11.empty())
+    for (int j = 0; j < n; ++j)
+      spline(_x1, _y[j], _y2_columns[j]);
+
+  else
+    for (int j = 0; j < n; ++j)
+      spline(_x1, _y[j], _y2_columns[j], _yx11[j], _yx1n[j]);
+
+}
+
+void
+BicubicSplineInterpolation::solve()
+{
+  constructRowSplineSecondDerivativeTable();
+  constructColumnSplineSecondDerivativeTable();
+}
+
+Real
+BicubicSplineInterpolation::sample(Real x1, Real x2, Real yx11/* = 1e30*/, Real yx1n/* = 1e30*/)
+{
+  int m = _x1.size();
+  std::vector<Real> column_spline_second_derivs(m), row_spline_eval(m);
+
+  // Evaluate m row-splines to get y-values for column spline construction
+  for (int j = 0; j < m; ++j)
+    row_spline_eval[j] = SplineInterpolationBase::sample(_x2, _y[j], _y2_rows[j], x2);
+
+  // Construct single column spline; get back the second derivatives wrt x1 coord on the x1 grid points
+  spline(_x1, row_spline_eval, column_spline_second_derivs, yx11, yx1n);
+
+  // Evaluate newly constructed column spline
+  return SplineInterpolationBase::sample(_x1, row_spline_eval, column_spline_second_derivs, x1);
+}
+
+Real
+BicubicSplineInterpolation::sampleDerivative(Real x1, Real x2, unsigned int deriv_var, Real yp1/* = 1e30*/, Real ypn/* = 1e30*/)
+{
+  // Take derivative along x1 axis
+  if (deriv_var == 1)
+  {
+    int m = _x1.size();
+    std::vector<Real> column_spline_second_derivs(m), row_spline_eval(m);
+
+    // Evaluate m row-splines to get y-values for column spline construction
+    for (int j = 0; j < m; ++j)
```

Comment:
Sadly, there is no construct that lets you loop over a range of integers like that. People have macros and other tricks for building them, but it's more trouble than it's worth IMHO.

Decltype lets you match types, so in the new C++11 way of thinking you use `auto` a lot and then to make sure you match one type against some other `auto` type, you use decltype to declare it. The downside is that it's a bit ugly.
```
for (decltype(m) j = 0; ...)
```